perception/image_projection_based_fusion/src/pointpainting_fusion/node.cpp

```diff
@@ -0,0 +1,253 @@
+  const std::string encoder_engine_path = this->declare_parameter("encoder_engine_path", "");
+  const std::string head_onnx_path = this->declare_parameter("head_onnx_path", "");
+  const std::string head_engine_path = this->declare_parameter("head_engine_path", "");
+  class_names_ = this->declare_parameter<std::vector<std::string>>("class_names");
+  rename_car_to_truck_and_bus_ = this->declare_parameter("rename_car_to_truck_and_bus", false);
+  has_twist_ = this->declare_parameter("has_twist", false);
+  const std::size_t point_feature_size =
+    static_cast<std::size_t>(this->declare_parameter<std::int64_t>("point_feature_size"));
+  const std::size_t max_voxel_size =
+    static_cast<std::size_t>(this->declare_parameter<std::int64_t>("max_voxel_size"));
+  const auto point_cloud_range = this->declare_parameter<std::vector<double>>("point_cloud_range");
+  const auto voxel_size = this->declare_parameter<std::vector<double>>("voxel_size");
+  const std::size_t downsample_factor =
+    static_cast<std::size_t>(this->declare_parameter<std::int64_t>("downsample_factor"));
+  const std::size_t encoder_in_feature_size =
+    static_cast<std::size_t>(this->declare_parameter<std::int64_t>("encoder_in_feature_size"));
+
+  centerpoint::NetworkParam encoder_param(encoder_onnx_path, encoder_engine_path, trt_precision);
+  centerpoint::NetworkParam head_param(head_onnx_path, head_engine_path, trt_precision);
+  centerpoint::DensificationParam densification_param(
+    densification_world_frame_id, densification_num_past_frames);
+  centerpoint::CenterPointConfig config(
+    class_names_.size(), point_feature_size, max_voxel_size, point_cloud_range, voxel_size,
+    downsample_factor, encoder_in_feature_size, score_threshold, circle_nms_dist_threshold);
+
+  // create detector
+  detector_ptr_ = std::make_unique<centerpoint::PointPaintingTRT>(
+    encoder_param, head_param, densification_param, config);
+
+  // sub and pub
+  sub_.subscribe(this, "~/input/pointcloud", rmw_qos_profile_sensor_data);
+  obj_pub_ptr_ = this->create_publisher<DetectedObjects>("~/output/objects", rclcpp::QoS{1});
+}
+
+void PointpaintingFusionNode::preprocess(sensor_msgs::msg::PointCloud2 & painted_pointcloud_msg)
+{
+  sensor_msgs::msg::PointCloud2 tmp;
+  tmp = painted_pointcloud_msg;
+
+  // set fields
+  sensor_msgs::PointCloud2Modifier pcd_modifier(painted_pointcloud_msg);
+  pcd_modifier.clear();
+  painted_pointcloud_msg.width = tmp.width;
+  pcd_modifier.setPointCloud2Fields(
+    7, "x", 1, sensor_msgs::msg::PointField::FLOAT32, "y", 1, sensor_msgs::msg::PointField::FLOAT32,
+    "z", 1, sensor_msgs::msg::PointField::FLOAT32, "intensity", 1,
+    sensor_msgs::msg::PointField::FLOAT32, "CAR", 1, sensor_msgs::msg::PointField::FLOAT32,
+    "PEDESTRIAN", 1, sensor_msgs::msg::PointField::FLOAT32, "BICYCLE", 1,
+    sensor_msgs::msg::PointField::FLOAT32);
+  painted_pointcloud_msg.point_step = 28;
```

Comment:
'28' is 7(num fields)x4(float32)?

```
  const int num_fields = 7;
   pcd_modifier.setPointCloud2Fields(
    7, "x", 1, sensor_msgs::msg::PointField::FLOAT32, "y", 1, sensor_msgs::msg::PointField::FLOAT32,
    "z", 1, sensor_msgs::msg::PointField::FLOAT32, "intensity", 1,
    sensor_msgs::msg::PointField::FLOAT32, "CAR", 1, sensor_msgs::msg::PointField::FLOAT32,
    "PEDESTRIAN", 1, sensor_msgs::msg::PointField::FLOAT32, "BICYCLE", 1,
    sensor_msgs::msg::PointField::FLOAT32);
  painted_pointcloud_msg.point_step = num_fields * sizeof(sensor_msgs::msg::PointField::FLOAT32);
```

Reply:
4 is the length of bytes of each field.

Reply:
I see.
How about this?
```
   constexpr int num_fields = 7;
   pcd_modifier.setPointCloud2Fields(
    num_fields, "x", 1, sensor_msgs::msg::PointField::FLOAT32, "y", 1, sensor_msgs::msg::PointField::FLOAT32,
    "z", 1, sensor_msgs::msg::PointField::FLOAT32, "intensity", 1,
    sensor_msgs::msg::PointField::FLOAT32, "CAR", 1, sensor_msgs::msg::PointField::FLOAT32,
    "PEDESTRIAN", 1, sensor_msgs::msg::PointField::FLOAT32, "BICYCLE", 1,
    sensor_msgs::msg::PointField::FLOAT32);
  painted_pointcloud_msg.point_step = num_fields * sizeof(float);
```

Reply:
I'd like to remove magic number like 7, 28

Reply:
Got it. Thanks for the advice, I have pushed the changes.

Reply:
![image](https://user-images.githubusercontent.com/8327598/173739977-6be9e04b-f558-4711-9709-2e30ac891f7b.png)
Still left `7` :pray:

Reply:
Sorry for the miss. I corrected it.